File: jtaf-spring/src/main/java/ch/jtaf/report/SeriesRanking.java
package ch.jtaf.report;

import ch.jtaf.entity.Athlete;
import ch.jtaf.entity.Competition;
import ch.jtaf.i18n.I18n;
import ch.jtaf.vo.SeriesRankingCategoryVO;
import ch.jtaf.vo.SeriesRankingVO;
import com.itextpdf.text.Document;
import com.itextpdf.text.DocumentException;
import com.itextpdf.text.PageSize;
import com.itextpdf.text.pdf.PdfPTable;
import com.itextpdf.text.pdf.PdfWriter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.Locale;

public class SeriesRanking extends Ranking {

    private final static Logger LOGGER = LoggerFactory.getLogger(SeriesRanking.class);

    private Document document;
    private final SeriesRankingVO ranking;

    public SeriesRanking(SeriesRankingVO ranking, Locale locale) {
        super(locale);
        this.ranking = ranking;
    }

    public byte[] create() {
        try {
            byte[] ba;
            try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
                document = new Document(PageSize.A4);
                PdfWriter pdfWriter = PdfWriter.getInstance(document, baos);
                pdfWriter.setPageEvent(new HeaderFooter(
                        I18n.getInstance().getString(locale, "Series Ranking"),
                        ranking.getSeries().getName(),
                        ""));
                document.open();
                createRanking();
                document.close();
                pdfWriter.flush();
                ba = baos.toByteArray();
            }
            return ba;
        } catch (DocumentException | IOException e) {
            LOGGER.error(e.getMessage(), e);
            return new byte[0];
        }
    }

    private void createRanking() throws DocumentException {
        for (SeriesRankingCategoryVO category : ranking.getCategories()) {
            if (numberOfRows > 22) {
                document.newPage();
            }
            PdfPTable table = createAthletesTable();
            createCategoryTitle(table, category);
            numberOfRows += 2;

            int position = 1;
            for (Athlete athlete : category.getAthletes()) {
                if (numberOfRows > 22) {
                    document.add(table);
                    document.newPage();
                    table = createAthletesTable();
                }
                createAthleteRow(table, position, athlete);
                position++;
                numberOfRows += 1;
            }
            document.add(table);
        }
    }

    private PdfPTable createAthletesTable() {
        PdfPTable table = new PdfPTable(new float[]{2f, 10f, 10f, 2f, 5f, 5f});
        table.setWidthPercentage(100);
        table.setSpacingBefore(cmToPixel(1f));
        return table;
    }

    private void createCategoryTitle(PdfPTable table, SeriesRankingCategoryVO category) {
        addCategoryTitleCellWithColspan(table, category.getCategory().getAbbreviation(), 1);
        addCategoryTitleCellWithColspan(table, category.getCategory().getName() + " "
                + category.getCategory().getYearFrom() + " - " + category.getCategory().getYearTo(), 5);

        addCategoryTitleCellWithColspan(table, " ", 6);
    }

    private void createAthleteRow(PdfPTable table, int position, Athlete athlete) {
        addCell(table, position + ".");
        addCell(table, athlete.getLastName());
        addCell(table, athlete.getFirstName());
        addCell(table, athlete.getYear() + "");
        addCell(table, athlete.getClub() == null ? "" : athlete.getClub().getAbbreviation());
        addCellAlignRight(table, athlete.getSeriesPoints(ranking.getSeries()) + "");

        StringBuilder sb = new StringBuilder();
        for (Competition competition : ranking.getSeries().getCompetitions()) {
            sb.append(competition.getName());
            sb.append(": ");
            sb.append(athlete.getTotalPoints(competition));
            sb.append(" ");
        }
        addCell(table, "");
        addResultsCell(table, sb.toString());
    }
}

File: src/main/kotlin/ch/jtaf/control/reporting/report/SeriesRanking.kt
package ch.jtaf.control.reporting.report

import ch.jtaf.control.reporting.data.SeriesRankingCategoryData
import ch.jtaf.control.reporting.data.SeriesRankingData
import ch.jtaf.entity.AthleteWithResultsDTO
import com.itextpdf.text.Document
import com.itextpdf.text.PageSize
import com.itextpdf.text.pdf.PdfPTable
import com.itextpdf.text.pdf.PdfWriter
import org.slf4j.LoggerFactory
import java.io.ByteArrayOutputStream
import java.util.*

class SeriesRanking(private val ranking: SeriesRankingData, locale: Locale) : Ranking(locale) {

    private val logger = LoggerFactory.getLogger(SeriesRanking::class.java)

    private var document: Document? = null

    fun create(): ByteArray {
        try {
            ByteArrayOutputStream().use { baos ->
                document = Document(PageSize.A4)
                val pdfWriter = PdfWriter.getInstance(document!!, baos)
                pdfWriter.pageEvent = HeaderFooter("Series Ranking", ranking.series.name, "")
                document!!.open()
                createRanking()
                document!!.close()
                pdfWriter.flush()
                return baos.toByteArray()
            }
        } catch (e: Exception) {
            logger.error(e.message, e)
            return ByteArray(0)
        }

    }

    private fun createRanking() {
        ranking.categories.forEach {
            if (numberOfRows > 22) {
                document!!.newPage()
            }
            var table = createAthletesTable()
            createCategoryTitle(table, it)
            numberOfRows += 2

            var position = 1
            it.getAthletesSortedByPointsDesc().forEach {
                if (numberOfRows > 22) {
                    document!!.add(table)
                    document!!.newPage()
                    table = createAthletesTable()
                }
                createAthleteRow(table, position, it)
                position++
                numberOfRows += 1
            }
            document!!.add(table)
        }
    }

    private fun createAthletesTable(): PdfPTable {
        val table = PdfPTable(floatArrayOf(2f, 10f, 10f, 2f, 5f, 5f))
        table.widthPercentage = 100f
        table.spacingBefore = cmToPixel(1f)
        return table
    }

    private fun createCategoryTitle(table: PdfPTable, category: SeriesRankingCategoryData) {
        addCategoryTitleCellWithColspan(table, category.category.abbreviation, 1)
        addCategoryTitleCellWithColspan(table, category.category.name + " " + category.category.yearFrom + " - " + category.category.yearTo, 5)
        addCategoryTitleCellWithColspan(table, " ", 6)
    }

    private fun createAthleteRow(table: PdfPTable, position: Int, athlete: AthleteWithResultsDTO) {
        addCell(table, position.toString() + ".")
        addCell(table, athlete.athlete.lastName)
        addCell(table, athlete.athlete.firstName)
        addCell(table, athlete.athlete.yearOfBirth.toString())
        addCell(table, if (athlete.athlete.club == null) "" else athlete.athlete.club!!.abbreviation)
        addCellAlignRight(table, athlete.results.sumBy { it.points }.toString())

        val sb = StringBuilder()
        ranking.series.competitions.forEach { competition ->
            sb.append(competition.name)
            sb.append(": ")
            sb.append(athlete.results.filter { it.competition == competition }.sumBy { it.points })
            sb.append(" ")
        }
        addCell(table, "")
        addResultsCell(table, sb.toString())
    }

}
